Fix analyticstack

itsjeyd opened the pull request:
cf. [OLIVE-5](https://openedx.atlassian.net/browse/OLIVE-5)

**Test instructions**

Testing this PR involves two steps:

1. Creating a base box using `vagrant/base/analyticstack/Vagrantfile`.
2. Setting up an instance of the Analytics Devstack based on the [official instructions](https://edx.readthedocs.org/projects/edx-installing-configuring-and-running/en/latest/devstack/analytics_devstack.html) using the newly created box file.

Detailed steps:

* Create the base box by doing:

  ```sh
  mkdir analytics
  cd analytics
  git clone https://github.com/open-craft/configuration.git
  cd configuration
  git checkout tim/fix-analyticstack
  virtualenv -p /usr/bin/python2 venv
  source venv/bin/activate
  pip install -r pre-requirements.txt
  pip install -r requirements.txt
  cd vagrant/base/analyticstack
  export OPENEDX_RELEASE=named-release/dogwood
  vagrant up
  ```

* Package the base box via:

  ```sh
  vagrant package
  ```

* Go back to parent directory of `configuration` repo and perform steps from "Install the Software Prerequisites" section of [Installing the Open edX Analytics Developer Stack](https://edx.readthedocs.org/projects/edx-installing-configuring-and-running/en/latest/devstack/analytics_devstack.html):

  ```sh
  deactivate
  cd ../../../..
  git clone https://github.com/edx/edx-analytics-pipeline
  cd edx-analytics-pipeline
  virtualenv -p /usr/bin/python2 venv
  source venv/bin/activate
  make bootstrap
  ```

* Create directory for Analytics Devstack and `cd` into it:

  ```sh
  deactivate
  cd ..
  mkdir analyticstack
  cd analyticstack
  ```

* Copy `vagrant/release/analyticstack/Vagrantfile` into this directory:

  ```sh
  cp ../configuration/vagrant/release/analyticstack/Vagrantfile ./
  ```

* Edit Vagrantfile to point to local box file:

  ```ruby
      Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
        # ...
        config.vm.box     = "package"
        config.vm.box_url = "file:///path/to/analytics/configuration/vagrant/base/analyticstack/package.box"
        # ...
  ```

* Use Vagrantfile to bring up a new instance of the Analytics Devstack:

  ```sh
  vagrant up
  ```

* Access the new instance and switch to the `master` branch of `edx-platform`:

  ```sh
  vagrant ssh
  sudo su edxapp
  git checkout master
  ```

  This is necessary because the Vagrantfile we used to bring up the instance in the previous step uses `configuration_version=master`, which [will no longer install Ruby](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/edx-code/25yctXaG4QE) as part of the `edxapp` role. The `release` branch of edx-platform currently still depends on Ruby for SASS compilation, so you'll run into errors if you try to, e.g., run the LMS while still on the `release` branch.

* Verify that you can run the LMS, the Data API, and Insights as described [here](https://edx.readthedocs.org/projects/edx-installing-configuring-and-running/en/latest/devstack/analytics_devstack.html#id16). To be able to run pipeline tasks via `remote-task`, you'll also need the changes from https://github.com/edx/edx-analytics-pipeline/pull/199.

**Related PRs**

- https://github.com/edx/edx-analytics-pipeline/pull/199

edx-webhook commented:
Thanks for the pull request, @itsjeyd! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fconfiguration&number=2753

mulby commented on playbooks/vagrant-analytics.yml:
```diff
@@ -24,7 +26,17 @@
+    - role: notifier
+      NOTIFIER_DIGEST_TASK_INTERVAL: "5"
+    - browsers
+    - browsermob-proxy
```
IIRC we deliberately removed notifier, browsers and browsermob-proxy

bradenmacdonald commented on playbooks/vagrant-analytics.yml:
```diff
@@ -24,8 +26,13 @@
     - elasticsearch
     - forum
+    - ecommerce
+    - ecomworker
```
Do we need ecommerce on the analytics devstack? Does analytics interact with ecommerce at all?

mulby replied:
There are some tasks that make use of the e-comm database, however, they are currently only used for internal reporting, not actually driving any Insights reports.

I would recommend leaving out these services for now.

bradenmacdonald commented on vagrant/base/analyticstack/Vagrantfile:
```diff
@@ -30,57 +19,34 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
+  config.vm.network :private_network, ip: vm_guest_ip
+
+  # If you want to run the box but don't need network ports, set VAGRANT_NO_PORTS=1.
+  # This is useful if you want to run more than one box at once.
```
Nice. Typically you'd have to use that together with `VAGRANT_GUEST_IP` to give each box a different IP, right?

itsjeyd replied:
@bradenmacdonald I think so, although I didn't try running and interacting with different services on two Devstacks simultaneously. If you just want to check something on another guest, you don't need the `VAGRANT_GUEST_IP` option.

bradenmacdonald commented on vagrant/base/analyticstack/Vagrantfile:
```diff
@@ -30,57 +19,34 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
+    config.vm.network :forwarded_port, guest: 8088, host: 8088  # Hadoop Resource Manager
+  end
 
-  config.vm.synced_folder  ".", "/vagrant", disabled: true
```
Is there a reason why you reinstated the `/vagrant` mount here in the base box?

itsjeyd replied:
@bradenmacdonald I initially removed this line because I noticed it had been removed from the Vagrantfile for the devstack base box as well, and I thought it would make sense to reduce the differences between these files as much as possible. How might we benefit from bringing it back?

bradenmacdonald replied:
@itsjeyd Last time I checked, if this line is not present to override the default, then vagrant creates a default NFS mount point at `/vagrant/`. We don't really use that for edX purposes, so that's why it's disabled on the release boxes. I don't suppose it really matters if it's present while creating the base box or not.

itsjeyd replied:
@bradenmacdonald OK thanks, I'll leave this change alone then, to stay consistent with the base Vagrantfile for devstack.

bradenmacdonald commented on vagrant/release/analyticstack/Vagrantfile:
```diff
@@ -1,4 +1,4 @@
-Vagrant.require_version ">= 1.5.3"
+Vagrant.require_version ">= 1.6.5"
```
Should you bump the version for the base Vagrantfile too?

itsjeyd replied:
@bradenmacdonald Nice catch, fixed in 779b15f2559fde3d55e68def152bba065461bc8f.

bradenmacdonald commented:
:+1: 

Nice work @itsjeyd! And thank you so much for the excellent testing instructions. It makes my life so much easier.

This was by far the most uneventful analytics devstack setup I've done, and the first time everything worked for me out of the box :)

The only issue I had was a known bug in vagrant (i.e. not a problem with your PR), where it gets its ssh keys mixed up when you have many different vagrant devstacks on the same host:
```
Bradens-MacBook-Pro:analytics-devstack braden$ vagrant ssh
vagrant@127.0.0.1's password:
```
Note that it's not supposed to prompt for a password, and this prevents ansible from working. But luckily it's possible to workaround, since the password is 'vagrant'

mulby commented:
:+1: This looks good to me, we should have someone from ops review it as well. Maybe @e0d?

itsjeyd commented:
@mulby @bradenmacdonald Thanks for reviewing! I addressed your comments.

It would probably be good to do another round of testing but I'll hold off on that for now in case there's additional feedback.

itsjeyd commented:
@mulby @eod @bradenmacdonald FYI, I went ahead and did another round of testing with the updated code and `OPENEDX_RELEASE` set to `named-releases/dogwood` instead of `named-releases/dogwood.rc` for base box provisioning; it went fine.

EoD commented:
Pardon me?

bradenmacdonald commented:
He meant @e0d not @eod, sorry :-/

EoD commented:
@bradenmacdonald haha. No worries.

edx-webhook commented:
Thanks for the pull request, @itsjeyd! I've created [OSPR-1154](https://openedx.atlassian.net/browse/OSPR-1154) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/configuration/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

itsjeyd commented:
@e0d Could you please let me know if you'll be able to review this PR? That would be great :)

nedbat commented:
I've deleted the OSPR ticket that wasn't supposed to be created.

bradenmacdonald commented:
Thanks @nedbat. Sorry for the noise.

e0d commented on playbooks/vagrant-analytics.yml:
```diff
@@ -24,8 +25,11 @@
     - elasticsearch
     - forum
+    - { role: 'rabbitmq', rabbitmq_ip: '127.0.0.1' }
+    - programs
```
I'm not sure why programs is included here, can you clarify?

itsjeyd replied:
@e0d When I set this up initially I thought it would be useful/convenient for development to have Analytics Devstack support a proper superset of the features that are included in Devstack out-of-the-box. That's why the `programs` role is listed here. However, to address other review notes I have been removing features that I had added initially (such as `ecommerce`), so there's actually no reason to keep just `programs`. Removed in 200fe08f3c76078b6cf8c373f71f884edc26a3cf.

e0d commented on vagrant/release/analyticstack/Vagrantfile:
```diff
@@ -17,24 +17,29 @@ export PYTHONUNBUFFERED=1
+# The vagrant-analytics.yml playbook will also do this, but only after loading playbooks into memory.
+# If these are out-of-date, this can cause problems
+# (e.g. permission issues, looking for templates that no longer exist).
+ansible-playbook -i localhost, -c local run_role.yml -e role=edx_ansible -e configuration_version=master
```
I realize this is perhaps beyond the scope of this PR, but doesn't overriding this to master here, given the use of named releases elsewhere, violate the principle of least astonishment?

itsjeyd replied:
@e0d 

> I realize this is perhaps beyond the scope of this PR, but doesn't overriding this to master here, given the use of named releases elsewhere, violate the principle of least astonishment?

It does. I modified the release Vagrantfile for analyticstack to support named releases. It now uses the same mechanism as the release Vagrantfile for devstack for updating the configuration repo. Aside from being less surprising, this should also further ease the process of evolving the two files in parallel. Let me know if you still think we should incorporate the ansible-bootstrap.sh script somehow. (Are there any plans for using it with devstack as well?)

itsjeyd commented:
@e0d Thanks for the feedback! I addressed your comments. Please have another look.

e0d commented:
:+1: from me.
